### PR TITLE
test: InfluxRPC frontend tag_values benchmarks

### PR DESCRIPTION
**Pull request by e-dard**
Contributing to https://github.com/influxdata/conductor/issues/159 through providing a framework for validating that performance optimisations are being correctly utilised on the InfluxRPC query path.

This PR provides some benchmarks of the `influxrpc` front-end, focussing only on `tag_values` initially.

## Goal

To have a place to run benchmarks at quite a high level where we can evaluate / verify the performance of certain operations on a chunk when that chunk is in different states. Specifically focussing on the performance differences between a chunk in the mutable buffer and a chunk(s) in the read buffer.

Finally, there is already a testing framework in then `server` crate that tests queries across chunks in different states; it would be beneficial to leverage that framework for the benchmarks too.

## Challenges/Choices

 - Exposing the `query_tests` module outside of the `server` crate. In the end I just kept the module private but I used the same trick I did in the read buffer, which was to re-export it as something public for `benchmarks` only. 
 - Exercising async code in criterion benchmarks. This actually wasn't too bad - it's documented [here](https://github.com/bheisler/criterion.rs/blob/master/book/src/user_guide/benchmarking_async.md).
 - Having a test fixture of line protocol that provides useful insight into how the performance of `tag_values` scales with cardinality of columns. For this I just used `inch` to generate some simple line-protocol with three tags varying in cardinality from 10–1000 values. I gzipped it up and put it in the test fixtures folder. It's 3.5MB, I hope folk aren't too bothered by that.
 - I needed to pick some combination of chunks to benchmark against. I decided to benchmark:
   * one MUB chunk;
   * one closed MUB chunk and one open MUB chunk;
   * one open MUB chunk and one Read Buffer chunk;
   * two Read Buffer chunks.
 - In all cases the same LP was loaded into each chunk. This means we can verify that optimisations for short-circuiting chunk execution are hooked up (hint, they're not yet).

## Benchmark Results

(I didn't bother with throughput but it can be added).

The following results show the performance of executing the equivalent of the following queries:

```
SHOW TAG VALUES WITH KEY = "tag{0, 1, 2}"; (`no_pred`)
SHOW TAG VALUES WITH KEY = "tag{0, 1, 2}" WHERE "tag2" = 'value321'; (`with_pred`)
```

```
⇒  critcmp master master                                                                                              group                                                                                master
-----                                                                                ------
Data in one open chunk and one closed chunk of mutable buffer/tag0/no_pred           1.00     91.0±2.55ms        ? ?/sec
Data in one open chunk and one closed chunk of mutable buffer/tag0/with_pred         1.00     11.5±0.72ms        ? ?/sec
Data in one open chunk and one closed chunk of mutable buffer/tag1/no_pred           1.00    120.3±5.10ms        ? ?/sec
Data in one open chunk and one closed chunk of mutable buffer/tag1/with_pred         1.00     11.2±0.22ms        ? ?/sec
Data in one open chunk and one closed chunk of mutable buffer/tag2/no_pred           1.00    203.2±8.45ms        ? ?/sec
Data in one open chunk and one closed chunk of mutable buffer/tag2/with_pred         1.00     11.2±0.21ms        ? ?/sec
Data in open chunk of mutable buffer, and one chunk of read buffer/tag0/no_pred      1.00    100.3±3.73ms        ? ?/sec
Data in open chunk of mutable buffer, and one chunk of read buffer/tag0/with_pred    1.00     31.2±1.80ms        ? ?/sec
Data in open chunk of mutable buffer, and one chunk of read buffer/tag1/no_pred      1.00    126.7±2.29ms        ? ?/sec
Data in open chunk of mutable buffer, and one chunk of read buffer/tag1/with_pred    1.00     33.0±1.70ms        ? ?/sec
Data in open chunk of mutable buffer, and one chunk of read buffer/tag2/no_pred      1.00    212.0±6.86ms        ? ?/sec
Data in open chunk of mutable buffer, and one chunk of read buffer/tag2/with_pred    1.00     18.1±0.99ms        ? ?/sec
Data in single open chunk of mutable buffer/tag0/no_pred                             1.00     98.7±6.08ms        ? ?/sec
Data in single open chunk of mutable buffer/tag0/with_pred                           1.00     11.2±0.37ms        ? ?/sec
Data in single open chunk of mutable buffer/tag1/no_pred                             1.00    118.9±3.97ms        ? ?/sec
Data in single open chunk of mutable buffer/tag1/with_pred                           1.00     11.7±0.64ms        ? ?/sec
Data in single open chunk of mutable buffer/tag2/no_pred                             1.00    202.1±8.49ms        ? ?/sec
Data in single open chunk of mutable buffer/tag2/with_pred                           1.00     11.1±0.27ms        ? ?/sec
Data in two read buffer chunks/tag0/no_pred                                          1.00    109.2±5.20ms        ? ?/sec
Data in two read buffer chunks/tag0/with_pred                                        1.00     44.2±1.83ms        ? ?/sec
Data in two read buffer chunks/tag1/no_pred                                          1.00    132.9±3.79ms        ? ?/sec
Data in two read buffer chunks/tag1/with_pred                                        1.00     41.7±2.43ms        ? ?/sec
Data in two read buffer chunks/tag2/no_pred                                          1.00    222.4±7.00ms        ? ?/sec
Data in two read buffer chunks/tag2/with_pred                                        1.00     27.9±0.92ms        ? ?/sec
``` 

## Benchmark Analysis

**Note**: During the development of these benchmarks the InfluxRPC frontend was updated in #1179 and other PRs, so the outcome of these results has changed.

Here is some insight into the above results:

~**Q**: Why are the Mutable Buffer benchmarks slower when using a predicate (`with_pred`) than not?~
~**A**: The Mutable Buffer doesn't currently support predicate pushdown, so instead all of the data is scanned out of the Mutable Buffer and send to DF for further processing.~

**Q**: Why are these queries slower when there is a chunk or chunks in the read buffer?
**A**: Currently the Read Buffer's dedication operation for this query is not wired up. That's the next PR. Right now
The query engine will fall back to scanning data out of the read buffer using the Read Buffer's `read_filter` operation, which supports project push-down and predicate push-down.

~**Q**: The Mutable Buffer is slower with a predicate than without, but the Read Buffer is fast with a predicate than without. Why?~
~**A**: The Read Buffer is faster with a predicate because it only scans out the matching rows (because of projection and predicate push-down). The MUB hasn't got predicate push-down wired up yet.~

The next PR will be wiring in the Read Buffer's optimised `tag_values` implementation - I expect to see a significant improvement on the benchmarks involving the Read Buffer.

### CODE REVIEW

**Comment by tustvold** [comment]
> I gzipped it up and put it in the test fixtures folder. It's 3.5MB, I hope folk aren't too bothered by that.

Having had past experiences of trying to remove large blobs from the git index, and the absolute nightmare it is, I wonder if we might avoid some future toil by using LFS for this?

File: server/benches/tag_values.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    executor: &Executor,
+    db: &Db,
+    tag_key: &str,
+    predicate: Predicate,
+) {
+    let plan = planner
+        .tag_values(db, &tag_key, predicate)
+        .await
+        .expect("built plan successfully");
+    let names = executor.to_string_set(plan).await.expect(
```
Comment (alamb) [inline server/benches/tag_values.rs:118]: FWIW after https://github.com/influxdata/influxdb_iox/pull/1191 can potentially run these on a separate async threadpool -- so when we integrate the two we'll probably need to pick a number of worker threads. Should it be 1?

**Review by alamb — APPROVED** [review]
I think this code looks good and makes sense to me 👍  The changes to the `query_test` framework look good too

I also tried running it locally which was cool. 

The results so far are fascinating. I think this suite will be a great source of data as we go and work on improving the performace

**Comment by e-dard** [comment]
@tustvold wrote:
> Having had past experiences of trying to remove large blobs from the git index, and the absolute nightmare it is, I wonder if we might avoid some future toil by using LFS for this?

I am super easy on what we do. Only thing I care about:

 - it doesn't complicate dev experience by having to install/use more tools;
 - a dev can run all tests/benchmarks on a train with no internet 

😄

File: internal_types/src/entry.rs [hunk 2]
```diff
@@ -1231,8 +1231,13 @@ pub mod test_helpers {
     use chrono::TimeZone;
     use influxdb_line_protocol::parse_lines;
 
-    /// Converts the line protocol to a vec of ShardedEntry with a single shard
-    /// and a single partition
+    // An appropriate maximum size for batches of LP to be written into IOx. Using
```
Comment (e-dard) [inline internal_types/src/entry.rs:1234]: @pauldix I modified this so there is an API to automatically split large LP files (`lp_to_entries`)

File: internal_types/src/entry.rs [hunk 3]
```diff
@@ -1243,6 +1248,24 @@ pub mod test_helpers {
             .entry
     }
 
+    /// Converts the line protocol to a collection of `Entry` with a single
+    /// shard and a single partition, which is useful for testing when `lp` is
+    /// large. Batches are sized according to LP_BATCH_SIZE.
+    pub fn lp_to_entries(lp: &str) -> Vec<Entry> {
```
Comment (pauldix) [inline internal_types/src/entry.rs:1254]: If this is a single shard it should just be `lp_to_entry` which already exists. The partitioner that uses is by the hour, but I see you're using that already.

File: server/src/db.rs [hunk 4]
```diff
@@ -756,11 +756,13 @@ impl CatalogProvider for Db {
 
 pub mod test_helpers {
     use super::*;
-    use internal_types::entry::test_helpers::lp_to_entry;
+    use internal_types::entry::test_helpers::lp_to_entries;
```
Comment (pauldix) [inline server/src/db.rs:759]: these changes seem unnecessary based on what you're trying to do. See previous comment on the definition of that function `lp_to_entries`

File: internal_types/src/entry.rs [hunk 5]
```diff
@@ -1243,6 +1248,24 @@ pub mod test_helpers {
...
+        let lines: Vec<_> = parse_lines(&lp).map(|l| l.unwrap()).collect();
+
+        lines
+            .chunks(LP_BATCH_SIZE)
+            .map(|batch| {
+                lines_to_sharded_entries(batch, sharder(1).as_ref(), &hour_partitioner())
```
Comment (pauldix) [inline internal_types/src/entry.rs:1260]: probably easier to just use `lp_to_entry`

File: server/benches/tag_values.rs (hunk 1 above)
Comment (e-dard) [reply server/benches/tag_values.rs:118]: I think 1 is probably simpler for benchmarks as it's a bit more predictable. We can tweak I guess.

**Review by domodwyer — APPROVED** [review]
I really enjoyed your self-hosted Q&A session 👍

**Pull request merged**